File: app/src/test/java/org/mifos/mobilebanking/RegistrationVerificationPresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.register.UserVerify;
import org.mifos.mobilebanking.presenters.RegistrationVerificationPresenter;
import org.mifos.mobilebanking.ui.views.RegistrationVerificationView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;
import okhttp3.ResponseBody;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by Sean Kelly on 9/12/17.
 */

@RunWith(MockitoJUnitRunner.class)
public class RegistrationVerificationPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    RegistrationVerificationView view;

    @Mock
    ResponseBody responseBody;

    private RegistrationVerificationPresenter presenter;
    private UserVerify userVerify;

    @Before
    public void setUp() throws Exception {
        presenter = new RegistrationVerificationPresenter(dataManager, context);
        presenter.attachView(view);

        userVerify = FakeRemoteDataSource.getUserVerify();
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testVerifyUser() throws Exception {
        when(dataManager.verifyUser(userVerify)).thenReturn(Observable.just(responseBody));

        presenter.verifyUser(userVerify);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showVerifiedSuccessfully();
        verify(view, never()).showError("");
    }

    @Test
    public void testVerifyUserFails() throws Exception {
        when(dataManager.verifyUser(userVerify)).thenReturn(Observable.<ResponseBody>error(new
                RuntimeException()));

        presenter.verifyUser(userVerify);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view, never()).showVerifiedSuccessfully();
        verify(view).showError("");
    }

}

File: app/src/test/java/org/mifos/mobile/RegistrationVerificationPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import okhttp3.ResponseBody

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.register.UserVerify
import org.mifos.mobile.presenters.RegistrationVerificationPresenter
import org.mifos.mobile.ui.views.RegistrationVerificationView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by Sean Kelly on 9/12/17.
 */
@RunWith(MockitoJUnitRunner::class)
class RegistrationVerificationPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: RegistrationVerificationView? = null

    @Mock
    var responseBody: ResponseBody? = null
    private var presenter: RegistrationVerificationPresenter? = null
    private var userVerify: UserVerify? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = RegistrationVerificationPresenter(dataManager!!, context!!)
        presenter?.attachView(view)
        userVerify = FakeRemoteDataSource.userVerify
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(Exception::class)
    fun testVerifyUser() {
        Mockito.`when`<Observable<ResponseBody?>?>(dataManager?.verifyUser(userVerify)).thenReturn(Observable.just(responseBody))
        presenter?.verifyUser(userVerify)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showVerifiedSuccessfully()
        Mockito.verify(view, Mockito.never())?.showError("")
    }

    @Test
    @Throws(Exception::class)
    fun testVerifyUserFails() {
        Mockito.`when`(dataManager?.verifyUser(userVerify)).thenReturn(Observable.error(RuntimeException()))
        presenter?.verifyUser(userVerify)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view, Mockito.never())?.showVerifiedSuccessfully()
        Mockito.verify(view)?.showError("")
    }
}